consensus/src/checkpoint.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::common::{Chain, ProtocolStateChain};
+use mina_crypto::base58::Base58Encodable;
+use mina_crypto::hash::{EpochSeed, StateHash};
+use mina_rs_base::protocol_state::ProtocolState;
+
+pub const SLOTS_PER_EPOCH: u32 = 7140;
+
+#[derive(Debug)]
+pub enum ConsensusErrTyp {
+    ConsensusInitFail,
+}
+pub fn init_checkpoints(genesis: &mut ProtocolState) -> Result<(), ConsensusErrTyp> {
+    genesis.body.consensus_state.staking_epoch_data.seed = EpochSeed::default();
+    genesis
+        .body
+        .consensus_state
+        .staking_epoch_data
+        .start_checkpoint = StateHash::default();
+    genesis
+        .body
+        .consensus_state
+        .staking_epoch_data
+        .lock_checkpoint = StateHash::default();
+    genesis
+        .body
+        .consensus_state
+        .staking_epoch_data
+        .epoch_length
+        .0 = 1;
+    genesis.body.consensus_state.next_epoch_data.seed =
+        Base58Encodable::from_base58("2vaRh7FQ5wSzmpFReF9gcRKjv48CcJvHs25aqb3SSZiPgHQBy5Dt")
+            .map_err(|_| ConsensusErrTyp::ConsensusInitFail)?;
+    genesis
+        .body
+        .consensus_state
+        .next_epoch_data
+        .start_checkpoint = StateHash::default();
+    genesis.body.consensus_state.next_epoch_data.lock_checkpoint =
+        Base58Encodable::from_base58("3NLoKn22eMnyQ7rxh5pxB6vBA3XhSAhhrf7akdqS6HbAKD14Dh1d")
+            .map_err(|_| ConsensusErrTyp::ConsensusInitFail)?;
+    genesis.body.consensus_state.next_epoch_data.epoch_length.0 = 2;
+    Ok(())
+}
+
+pub fn is_short_range(
+    c0: &ProtocolStateChain,
+    c1: &ProtocolStateChain,
+) -> Result<bool, ConsensusErrTyp> {
+    let s0 = &c0
+        .consensus_state()
+        .ok_or(ConsensusErrTyp::ConsensusInitFail)?;
+    let s1 = &c1
+        .consensus_state()
+        .ok_or(ConsensusErrTyp::ConsensusInitFail)?;
+    let s0_lock_checkpoint = &s0.staking_epoch_data.lock_checkpoint;
+    let s1_lock_checkpoint = &s1.staking_epoch_data.lock_checkpoint;
+    let s1_next_epoch_lock_checkpoint = &s1.next_epoch_data.lock_checkpoint;
+
+    if s0.epoch_count == s1.epoch_count {
+        return Ok(s0_lock_checkpoint == s1_lock_checkpoint)
+    }
+
+    if s0.epoch_count.0 == s1.epoch_count.0 + 1
+        && Chain::epoch_slot(c1) >= Some(SLOTS_PER_EPOCH * 2 / 3)
```

Comment:
instead of `Chain::epoch_slot`, we can do `c1.epoch_slot()`

Reply:
Thank you @creativcoder 😄